### PR TITLE
Implement "Go Last"

### PR DESCRIPTION
**Pull request by brandonbloom**
This implements a "Go Last" command. This command will navigate either forward or backwards in history to the last entry in the history stack you were looking at. I've implemented this in service of implementing `''` and <code>``</code> for vim mode: https://github.com/VSCodeVim/Vim/issues/1089

### CODE REVIEW

**Comment by bpasero** [comment]
@brandonbloom trying to understand the purpose of the command:

> This command will navigate either forward or backwards in history to the last entry in the history stack you were looking at.

Is the idea that you just always remember the index of the last history operation by the user (either going forward or backward) so that you can jump back to that index? It is not clear to me how it would ever jump forward because once you go back and then navigate to another editor, the future history is lost.

**Comment by brandonbloom** [comment]
My goal is to implement a pair of vim commands that I use regularly: https://github.com/VSCodeVim/Vim/pull/1993

I typically use the line-wise `''`, but it can be defined in terms of the character-wise <code>``</code> operation. The VIM docs has this to say about `'` and <code>`</code>:

```
Jumping to a mark can be done in two ways:
1. With ` (backtick):	  The cursor is positioned at the specified location
			  and the motion is |exclusive|.
2. With ' (single quote): The cursor is positioned on the first non-blank
			  character in the line of the specified location and
			  the motion is linewise.
```

And this to say about the `''` and <code>``</code> chords:

```
							*''* *``*
''  ``			To the position before the latest jump, or where the
			last "m'" or "m`" command was given.  Not set when the
			|:keepjumps| command modifier was used.
			Also see |restore-position|.
```

In these terms, both any cursor motion, including navigating forwards or backwards, counts as a "jump".

You jump somewhere, and then tap `''` as many times as you like to flip back and forth between the place you came from and the place you went to. If you only execute the command once, it's no different than `<C-o>`, which is the "Navigate Back" command. If you execute it multiple times, it is essentially the same as alternating between "Navigate Back" and "Navigate Forward". This is extremely useful for quickly comparing two places within a single file.

VSCodeVim does not currently maintain its own "jump list", it relies on Code's navigation history. I implemented the simplest subset of the functionality I could think of, in the lowest risk way I could think of. That is: I just maintain the last place in history that you were. If there's any chance that last place is invalidated (such as when truncating history), you probably navigated somewhere, so "Navigate Back" is a reasonable thing to do.

Does this explanation clear things up?

**Comment by bpasero** [comment]
> You jump somewhere, and then tap '' as many times as you like to flip back and forth between the place you came from and the place you went to. 

Isn't the same possible by using the Ctrl+Tab keys and releasing to jump between MRU editors?

panel-red

**Comment by brandonbloom** [comment]
1) No. That's not the same thing for a few reasons. Most importantly, that won't work on one file in one editor. This is common when fiddling with imports while working, or when referring to some types at the top of a file, for example.

2) The goal is to provide even a crude approximation of a pair of very commonly used vim bindings.

2)

File: src/vs/workbench/browser/parts/editor/editor.contribution.ts [hunk 1]
```diff
@@ -368,6 +368,7 @@ registry.registerWorkbenchAction(new SyncActionDescriptor(FocusPreviousGroup, Fo
 registry.registerWorkbenchAction(new SyncActionDescriptor(FocusNextGroup, FocusNextGroup.ID, FocusNextGroup.LABEL, { primary: KeyChord(KeyMod.CtrlCmd | KeyCode.KEY_K, KeyMod.CtrlCmd | KeyCode.RightArrow) }), 'View: Focus Next Group', category);
 registry.registerWorkbenchAction(new SyncActionDescriptor(NavigateForwardAction, NavigateForwardAction.ID, NavigateForwardAction.LABEL, { primary: null, win: { primary: KeyMod.Alt | KeyCode.RightArrow }, mac: { primary: KeyMod.WinCtrl | KeyMod.Shift | KeyCode.US_MINUS }, linux: { primary: KeyMod.CtrlCmd | KeyMod.Shift | KeyCode.US_MINUS } }), 'Go Forward');
 registry.registerWorkbenchAction(new SyncActionDescriptor(NavigateBackwardsAction, NavigateBackwardsAction.ID, NavigateBackwardsAction.LABEL, { primary: null, win: { primary: KeyMod.Alt | KeyCode.LeftArrow }, mac: { primary: KeyMod.WinCtrl | KeyCode.US_MINUS }, linux: { primary: KeyMod.CtrlCmd | KeyMod.Alt | KeyCode.US_MINUS } }), 'Go Back');
+registry.registerWorkbenchAction(new SyncActionDescriptor(NavigateLastAction, NavigateLastAction.ID, NavigateLastAction.LABEL, { primary: null, win: null, mac: null, linux: null }), 'Go Last');
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/editor/editor.contribution.ts]: Can be written as:

`registry.registerWorkbenchAction(new SyncActionDescriptor(NavigateLastAction, NavigateLastAction.ID, NavigateLastAction.LABEL), 'Go Last');`

**Review by bpasero — CHANGES_REQUESTED** [review]

**Comment by bpasero** [comment]
@brandonbloom can you also create an issue in VSCode repo so that we can resolve it with this PR. The issue best explains this feature so that it can be tested during end game (e.g. what is the expected behaviour).

File: src/vs/workbench/browser/parts/editor/editor.contribution.ts (hunk 1 above)
Comment (brandonbloom) [reply src/vs/workbench/browser/parts/editor/editor.contribution.ts]: Fixed.

**Comment by brandonbloom** [comment]
Done. See https://github.com/Microsoft/vscode/issues/33902

**Comment by bpasero** [comment]
Thanks 👍

**Pull request merged**